### PR TITLE
[Fixes #10290] complete ISO contact roles per ressource base with multiplicity

### LINKED ISSUE #10290: DRAFT: Multiple contact roles per ressource base
**THIS IS A DRAFT - I WILL CONTINUE TO ELABORATE ON THIS LATER TODAY**

**Is your feature request related to a problem? Please describe.**
I want to use multiple CI CONTACT roles in the metadata of my datasets. 

a) I can not add multiple CI CONTACT roles to a dataset in the metadata wizard / advanced metadata editor.
b) I can not use all of the CI CONTACT roles ISO 19113 provides.
c) pycsw does not respond with all CI CONTACT roles given. It will only respond with POC ('pointOfContact'), the owner ('originator') and metadata author ('author'). 

I can put 1:N metadata roles in the Django admin area. Since Contact Roles are already implemented in the model:
https://github.com/GeoNode/geonode/blob/master/geonode/base/models.py#L105
which uses this enum with all ISO CI CONTACT roles: https://github.com/GeoNode/geonode/blob/6e093976049b6bff7d7af70f76df14e0f2f40e0a/geonode/people/enumerations.py#L21

but currently, you can only set POC and metadata author and owner by the metadata wizard. Some example of this is:
https://github.com/GeoNode/geonode/blob/6e093976049b6bff7d7af70f76df14e0f2f40e0a/geonode/base/models.py#L1850

**Describe the solution you'd like**
While POC, owner and metadata author are single contacts, ISO also provides roles which can have multiple persons involved.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by mwallschlaeger**
<Include a few sentences describing the overall goals for this Pull Request>
This is a work in progress PR to give an overview and first feedback from as the changes getting a little out of hand. 

What the current state is aware of:
- showing all contact roles via the rest API v2
- handling contact roles in metadata editor for dataset, I guess i would fully adopt this the same way for documents, maps and geoapps. I started working on this already (see changes).
- adopted resource base model to handle all contact roles and be able to set multiple authors for one role 

I know that further work in this topic at least requires changes to the following sections/files:

There is even more files to change apart from the ones already changed an the ones listed above. Further tests have to be added to this PR.

I'm willing to further work on this issue the upcoming weeks, but I think it is good to get some advice from the PSC to see if the direction this is going fits the ideas of the committee members.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/GeoNode/geonode/blob/master/CONTRIBUTING.md) 
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in the documentation)
- [x] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.

The following are required only for core and extension modules (they are welcomed, but not required, for contrib modules):
- [x] There is a ticket in https://github.com/GeoNode/geonode/issues describing the issue/improvement/feature (a notable exemption is, changes not visible to end-users)
- [x] The issue connected to the PR must have Labels and Milestone assigned
- [ ] PR for bug fixes and small new features are presented as a single commit
- [x] Commit message must be in the form "[Fixes #<issue_number>] Title of the Issue"
- [x] New unit tests have been added covering the changes, unless there is an explanation on why the tests are not necessary/implemented
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] This PR passes the QA checks: flake8 geonode
- [x] Commits changing the **settings**, **UI**, **existing user workflows**, or adding **new functionality**, need to include documentation updates
- [x] Commits adding **new texts** do use gettext and have updated .po / .mo files (without location infos)

**Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.**

### CODE REVIEW

**Comment by MalteIwanicki** [comment]
[This PR](https://github.com/GeoNode/geonode-mapstore-client/pull/1360) also needs to be merged when we wanna use a list instead of a single contact for point of contact.

**Comment by MalteIwanicki** [comment]
I took care of the contacts in the `geonode_metadata_full.html`. Once I can commit to your branch it's there

**Comment by gannebamm** [comment]
@MalteIwanicki @mwallschlaeger please take a peak at https://github.com/GeoNode/geonode/issues/10521
I think the mentioned changes could improve this draft PR

**Comment by gannebamm** [comment]
@mwallschlaeger Can we get a status report on this draft? What is missing. do you need help?

**Comment by mwallschlaeger** [comment]
Hi,

yes actually I'm waiting for @ahmdthr for a final commit to this PR, so we could leave the draft state. But i haven't heard from him in a while

**Comment by gannebamm** [comment]
@mwallschlaeger is this ready for review?

**Comment by mwallschlaeger** [comment]
@gannebamm yes it would be very good if smb. would check this PR and gives us feedback about further required changes

**Comment by giohappy** [comment]
@mwallschlaeger we will make some tests on our side. Overall it looks quite good so far. There are few places where we might add small comments.

**Review by giohappy — COMMENTED** [review]
@mwallschlaeger after adding multiple Points of Contacts they're listed inside the metadata detail but the editing field (both inside the wizard and the advacned page) only shows the first one.

File: geonode/base/models.py [hunk 1]
```diff
@@ -1644,45 +1608,232 @@ def maintenance_frequency_title(self):
...
+    def __get_poc__(self) -> List[settings.AUTH_USER_MODEL]:
+        return self.__get_contact_role_elements__(role="pointOfContact")
+
+    def __set_poc__(self, user_profile):
+        return self.__set_contact_role_element__(user_profile=user_profile, role="pointOfContact")
+
+    poc = property(__get_poc__, __set_poc__)
+
+    @property
+    def poc_csv(self):
```
Comment (mattiagiupponi) [inline geonode/base/models.py:1726]: I'll evaluate changing the name on `poc_as_csv` for all the new configurations added to make it more explicit

File: geonode/geoserver/helpers.py [hunk 2]
```diff
@@ -2083,14 +2083,14 @@ def sync_instance_with_geoserver(instance_id, *args, **kwargs):
...
-                            "title": str(instance.poc),
+                            "title": str(instance.poc_csv),
                             "width": None,
                             "height": None,
                             "href": None,
                             "url": None,
                             "type": None,
                         }
-                        profile = get_user_model().objects.get(username=instance.poc.username)
+                        profile = get_user_model().objects.get(username=instance.poc[0].username)
```
[the remark this answers is not in the document]
Comment (mattiagiupponi) [reply geonode/geoserver/helpers.py:2093]: I agree, access directly to the list index could be error prone

File: geonode/locale/it/LC_MESSAGES/django.po [hunk 3]
```diff
@@ -1251,6 +1251,12 @@ msgstr "Altri metadati opzionali"
 msgid "Responsible Parties"
 msgstr "Parti responsabili"
 
+msgid "toggle more Contact Roles"
+msgstr "mostrare più ruoli"
```
Comment (mattiagiupponi) [inline geonode/locale/it/LC_MESSAGES/django.po:1255]: You can change this in `mostra più ruoli`

File: geonode/maps/templates/layouts/map_panels.html [hunk 4]
```diff
@@ -543,32 +541,109 @@
...
-                          <span><label for="{{ map_form.metadata_author|id }}">{{ map_form.metadata_author.label }}</label></span>
-                          <!--<p class="xxs-font-size">(Author of the metadata)</p>-->
-                          {{ map_form.metadata_author }}
-                      </div>
+                    </div>
+                  {% endblock map_owner %}
+                </div>
+              </div>
+            <a href="#id-more-metadata-panel" data-toggle="collapse">{% trans "toggle more Contact Roles" %}</a>
+            <div class="panel panel-default panel-collapse collapse" collapsed id='id-more-metadata-panel'>
```
Comment (mattiagiupponi) [inline geonode/maps/templates/layouts/map_panels.html:567]: To reduce the verbosity in the HTML cannot we have a simple loop based on an enum where we can define the names and use it to generate the HTML??

File: geonode/layers/views.py [hunk 5]
```diff
@@ -578,11 +541,9 @@ def dataset_metadata(
             la.featureinfo_type = form["featureinfo_type"]
             la.save()
 
-        if new_poc is not None or new_author is not None:
-            if new_poc is not None:
-                layer.poc = new_poc
-            if new_author is not None:
-                layer.metadata_author = new_author
+        # update contact roles
+        layer.set_contact_roles_from_metadata_edit(dataset_form)
```
Comment (mattiagiupponi) [inline geonode/layers/views.py:545]: I see that the method `set_contact_roles_from_metadata_edit` return a `failed=True` if something went wrong.
If for any reason is not possible to set the contact roles, isn't it better to make the save fail?
even something like:

```python
is_failed = layer.set_contact_roles_from_metadata_edit(dataset_form)
if is_failed:
    raise Exception("Error during the contact role assignment")
```

**Review by mattiagiupponi — DISMISSED** [review]
Hi @mwallschlaeger 
Follows some first impression about the PR, since is huge it will take some time to review it properly

**Comment by ridoo** [comment]
> [This PR](https://github.com/GeoNode/geonode-mapstore-client/pull/1360) also needs to be merged when we wanna use a list instead of a single contact for point of contact.

That PR is outdated (and even fails to start). However, GMSC has a refactored the DetailsViewer component now and POCs are configurable in the `localConfig.json` now. Looping over a list of POCs should be done [in the plugin context utility](https://github.com/GeoNode/geonode-mapstore-client/blob/master/geonode_mapstore_client/client/js/utils/PluginsContextUtils.js#L24) .

**Comment by ridoo** [comment]
PR https://github.com/GeoNode/geonode-mapstore-client/pull/1421 supersedes PR https://github.com/GeoNode/geonode-mapstore-client/pull/1360 and a respects recent refactoring of  `DetailInfos`.

**Comment by giohappy** [comment]
@ridoo can you please take a look at the conflicts?

**Comment by ridoo** [comment]
@giohappy fixed it .. please have a look at my comment https://github.com/GeoNode/geonode-mapstore-client/pull/1421#issuecomment-1561350515

**Comment by mwallschlaeger** [comment]
I will do some checks again on this PR related to the rest-api and than reopen it again for a review.

**Comment by mwallschlaeger** [comment]
ok, fixed an issue related to the API, but now I think it's ready for a review again. I used https://github.com/GeoNodeUserGroup-DE/geonodectl to set contact roles via API using something like: `./geonodectl ds patch 1 --set '{"processor": [{ "pk": 1000, "username": "testuser" }, { "pk": 1001 }]}'`

**Comment by mwallschlaeger** [comment]
i will check them

**Comment by mwallschlaeger** [comment]
@giohappy :-) Could you give me another review on this PR so we can bring it to an end?^^

**Comment by giohappy** [comment]
Hi @mwallschlaeger. Unfortunately, there are conflicts again since a few changes have been committed recently to the metadata editor templates. They should be easy to fix.

We took another look at the PR. We don't see anything that blocks its merge, although we would have some remarks, in particular on the choice of showing all the possible roles inside the editor, instead of giving the choice to add them dynamically (like with a plus button). 

Hopefully, the metadata editor will be rewritten, and this new functionality will have to be taken into account for the new version.

**Comment by giohappy** [comment]
@mwallschlaeger please take a look at the failing test

**Comment by mwallschlaeger** [comment]
@giohappy review pls

**Comment by giohappy** [comment]
@mwallschlaeger I'm testing it right now. I can see the new widgets and I'm able to select multiple users, but they're not saved to the DB. both the metadata editor and the API keep showing a single user.

**Comment by mwallschlaeger** [comment]
@giohappy i will check this again. but what I recognized when I applied those changes to an existing Geonode Instance is that there are migrations required for the already uploaded resources

**Comment by giohappy** [comment]
>@giohappy i will check this again. but what I recognized when I applied those changes to an existing Geonode Instance is that there are migrations required for the already uploaded resources

which migrations @mwallschlaeger? I don'ìt see migrations in your PR

**Comment by mwallschlaeger** [comment]
> > @giohappy i will check this again. but what I recognized when I applied those changes to an existing Geonode Instance is that there are migrations required for the already uploaded resources
> 
> which migrations @mwallschlaeger? I don'ìt see migrations in your PR

Hey the migrations are not included yet. but I tested this again they are not required. Regarding your issue. I cannot confirm that multiple users are not set correctly. Actually there is also a test for this case in `geonode.layers.tests.test_resource_form_is_valid_multiple_user_contact_role_as_queryset` and `geonode.layers.tests.test_patch_processor` and so on are tests setting multiple users for one contact role.

Can you please doublecheck that you have the right commit?

**Comment by giohappy** [comment]
@mwallschlaeger ok, I will double check, BTW your latest merges seem to have broken the tests again.

**Comment by giohappy** [comment]
@MalteIwanicki I confirm it works as expected now. I think this can be merged as soon as the tests are fixes.

**Comment by MalteIwanicki** [comment]
@giohappy @ridoo  I'm totally out of the loop for this PR, not knowing what it was about. It has been over a year. Do you need any action from my side?

**Comment by ridoo** [comment]
> @giohappy @ridoo I'm totally out of the loop for this PR, not knowing what it was about. It has been over a year. Do you need any action from my side?

@MalteIwanicki no problem at all :)

File: geonode/base/models.py [hunk 6]
```diff
@@ -545,8 +517,6 @@ class ThesaurusKeyword(models.Model):
 
     thesaurus = models.ForeignKey("Thesaurus", related_name="thesaurus", on_delete=models.CASCADE)
 
-    image = models.CharField(max_length=512, help_text="A URL to an image", null=True, blank=True)
```
Comment (giohappy) [inline geonode/base/models.py:548]: @mwallschlaeger why did you drop this field? It was introduced in https://github.com/GeoNode/geonode/commit/c437a0a69054217976fa519b245e82966543dbe5

**Review by giohappy — CHANGES_REQUESTED** [review]

File: geonode/base/models.py (hunk 6 above)
Comment (etj) [inline geonode/base/models.py:548]: This field has been removed

**Review by giohappy — APPROVED** [review]

**Comment by giohappy** [comment]
@mwallschlaeger FYI we're investigating several problems that were reported with the management of contacts in metadata. It's broken in many ways. We will open an issue soon but I was curious to know if you did notice the problems it's giving, like:

 - contacts are not saved correctly or they're not saved at all
 - the layout for documents is broken

**Comment by mwallschlaeger** [comment]
@giohappy hey i have noticed one issue I have fixed in our fork. i can at least play the one I have fixed back to core in the next days

**Comment by giohappy** [comment]
@mwallschlaeger do you mean that editing of contacts works fine (except the issue you mentioned)?
If you try on the [stable demo](https://stable.demo.geonode.org/) you'll see that the changes to contacts are not saved.

**Comment by giohappy** [comment]
@mwallschlaeger I've seen your PR but tests are failing. You're using `getlist` on a dict object.
I wonder how much your fork has diverged from this.

If it cannot be easily fixed we will be forced to revert the commit the commits. I have built a Docker image at the commit of your PR and it seems they weren't working even at that time. Something must be changed between my tests of your PR and its merge.

**Comment by gannebamm** [comment]
@vineetasharma105 

Please check this in our dev instance.

**Comment by gannebamm** [comment]
@giohappy If this is not easily fixable, than my +1 for reverting the changes for now.

**Comment by mwallschlaeger** [comment]
@giohappy I guess i figured it out now. The test which are failing are providing the form data using the data argument, which passes and empty dict to the not filled contact roles, where `getlist` cannot be executed on. 

As in e.g.
https://github.com/GeoNode/geonode/blob/75edcdb50417f96d858b25b46a0998db6fe2ec14/geonode/layers/views.py#L532

they get filled via `initial`.

Another problem I came across investigating this, was that contact roles couldn't be removed also this was related to the change of the expected object type in widgets.py.

I'm gonna run the tests now locally and commit the changes to the PR, so you can check it, too.

**Comment by giohappy** [comment]
@mwallschlaeger noticed that we see different behavior between stable demo (4.2.x) and master demo. 

The latter is even more worrying because changes are applied when a user is selected for a role, but the resulting saved user is different from the one that was selected.

**Pull request merged**